### DISCUSSION TITLE
Tyro79 - i2c2 works on betaflight but not inav

### DISCUSSION THREAD

**Discussion by vcrparts** [iNavFlight/inav#6716] [Q&A]
I have a Tyro79 with an [Eachines STRM32F411ce based flight controller](https://usa.banggood.com/Eachine-F4-F411-Flight-Controller-for-LAL3-145mm-3-Inch-FPV-Racing-Drone-OSD-5V-BEC-20x20mm-p-1693538.html).  It is compatible with the MatekF411 target however the i2c1 pins are not exposed on the board.  I would like to add a BMP280 barometer for altitude hold and have tried editing and compiling a custom target iNav firmware.  [According to the datasheet](https://www.st.com/resource/en/datasheet/stm32f411ce.pdf) I should be able to use the S5 pad from PB03 as i2c2 SDA; and the S6 pad from PB10 as i2c2 SCL.  

Additionally, I'd like to use the LED output on PA8 as a servo output, but lets worry about the i2c for now.

I know my wiring and port numbers are correct as I've been successful in getting the barometer to work on i2c2 using these pins under betaflight using the stock MatekF411 firmware and assigning the pins using betaflight's resource mapping.

Attached CLI output from betaflight showing steps for success with barometer:
[BTFL working barometer steps.txt](https://github.com/iNavFlight/inav/files/6153782/BTFL.working.barometer.steps.txt)
Just four steps in betaflight:
resource I2C_SCL 2 B10
resource I2C_SDA 2 B03
set baro_i2c_device = 2
set baro_hardware = BMP280

In iNav, barometer auto does not detect any barometer.  After setting barometer to BMP280, barometer icon is red, and barometer = unavailable in cli status.  Inav shows 10 or 55 i2c errors depending if I comment out i2c1 settings.  i2c speeds make no difference.

Here's what I've tried to do, modifying target.c and target.h for the MATEKF411.  Is there something else I need to do to enable i2c2?

Please see attached files:
inav/src/main/target/MATEKF411/target.c 
[target.c.txt](https://github.com/iNavFlight/inav/files/6153759/target.c.txt)
inav/src/main/target/MATEKF411/target.h
[target.h.txt](https://github.com/iNavFlight/inav/files/6153760/target.h.txt)

iNav resource and dump:
[INAV_cli_status resource and dump.txt](https://github.com/iNavFlight/inav/files/6153795/INAV_cli_status.resource.and.dump.txt)

Thanks in advance for any help you can offer.

Also wanted to say thanks to all those who've done so much work to make iNav.  Amazing!  And special thanks to Paweł Spychalski, I've learned so much from your videos.

**Comment by vcrparts** [comment]
I've spent a couple weeks worth of free time trying different variations of editing and compiling the source code.  After considering the time I've spent, I've gone ahead and ordered a FC with a built in barometer.

I still feel this is a useful endeavor to get the barometer working with iNav on the tyro79.  From what I can tell the Eachine Tyro79 is popular, in particular with people new to quad-copters who would most benefit from the option of altitude hold.

So any suggestions are welcome.

Here's some questions I have that I couldn't answer from looking at the source code: 

1)  Alternate Functions -  My setup uses i2c alternate functions from different Alternate Function columns:   AF09 for SDA and AF04 for SCL.   Could the be an issue for me???  
![strm32f411](https://user-images.githubusercontent.com/80803229/111845478-e9aef880-88d2-11eb-9fb4-4641e2607a23.jpg)

lib/main/STM32F4/Drivers/STM32F4xx_StdPeriph_Driver/inc/stm32f4xx_gpio.h
and 
 inav/lib/main/STM32F4/Drivers/STM32F4xx_StdPeriph_Driver/inc/stm32f4xx_gpio.h 
Look like they may assume i2c2 have SDA and SCL on the same Alternate Function configuration
I'll try calling these commands explicitly outside of the if/else build logic:
#define GPIO_AF_I2C2          ((uint8_t)0x04)  /* I2C2 Alternate Function mapping */
#define GPIO_AF9_I2C2         ((uint8_t)0x09)  /* I2C2 Alternate Function mapping (Only for STM32F401xx/STM32F410xx/STM32F411xE/STM32F412xG Devices) */

2) pullup.... I just saw that some targets have a "USE_I2C_PULLUP".  I'll try enabling that...

3) is there an easy way to confirm sections of source code have run?  Writeline to console/log/blackbox?

Thanks again!

**Reply by WladimirWW** [reply]
Yes, you are right.
'Inav' i2cInit() function use AF04 for STM32F4 :
(inav/src/main/drivers/bus_i2c_stm32f40x.c line >485)



```
...
    IOInit(sda, OWNER_I2C, RESOURCE_I2C_SDA, RESOURCE_INDEX(device));

#ifdef STM32F4
    IOConfigGPIOAF(scl, IOCFG_I2C, GPIO_AF_I2C);
    IOConfigGPIOAF(sda, IOCFG_I2C, GPIO_AF_I2C); // <-- wrong alternate function for PB3 as i2c2 SDA
#pragma message " SDA = " sda
#else
    IOConfigGPIO(scl, IOCFG_I2C);
    IOConfigGPIO(sda, IOCFG_I2C);
#endif
....
```
replace with:
```
...
    IOInit(sda, OWNER_I2C, RESOURCE_I2C_SDA, RESOURCE_INDEX(device));

#ifdef STM32F4
    IOConfigGPIOAF(scl, IOCFG_I2C, GPIO_AF_I2C);
    IOConfigGPIOAF(sda, IOCFG_I2C, GPIO_AF9_I2C2); // <-- right alternate function for PB3 as i2c2 SDA
#pragma message " SDA = " sda
#else
    IOConfigGPIO(scl, IOCFG_I2C);
    IOConfigGPIO(sda, IOCFG_I2C);
#endif
...
```
My tyro79 work fine with i2c2 baro+mag+ usart1 gps
[fixed.tar.gz](https://github.com/iNavFlight/inav/files/6481114/fixed.tar.gz)
